File: internal/handlers/imme.go
package handlers

import (
	"context"
	"fmt"
	"net/http"
	"time"

	"github.com/gin-gonic/gin"
	"github.com/xxnuo/MTranServer/internal/services"
)

var immeLangToBCP47 = map[string]string{
	"zh-CN": "zh-Hans",
	"zh-TW": "zh-Hant",
}

func convertImmeLangToBCP47(immeLang string) string {

	if bcp47, ok := immeLangToBCP47[immeLang]; ok {
		return bcp47
	}

	return immeLang
}

type ImmeTranslateRequest struct {
	SourceLang string   `json:"source_lang" binding:"required" example:"en"`
	TargetLang string   `json:"target_lang" binding:"required" example:"zh-CN"`
	TextList   []string `json:"text_list" binding:"required" example:"Hello, world!,Good morning!"`
}

type ImmeTranslation struct {
	DetectedSourceLang string `json:"detected_source_lang" example:"en"`
	Text               string `json:"text" example:"你好，世界！"`
}

type ImmeTranslateResponse struct {
	Translations []ImmeTranslation `json:"translations"`
}

// HandleImmeTranslate 沉浸式翻译插件接口
// @Summary      沉浸式翻译插件接口
// @Description  为沉浸式翻译插件提供的翻译接口
// @Tags         插件
// @Accept       json
// @Produce      json
// @Param        token    query     string                  false  "API Token"
// @Param        request  body      ImmeTranslateRequest    true   "沉浸式翻译请求"
// @Success      200      {object}  ImmeTranslateResponse
// @Failure      400      {object}  map[string]string
// @Failure      401      {object}  map[string]string
// @Failure      500      {object}  map[string]string
// @Router       /imme [post]
func HandleImmeTranslate(apiToken string) gin.HandlerFunc {
	return func(c *gin.Context) {

		if apiToken != "" {
			token := c.Query("token")
			if token != apiToken {
				c.JSON(http.StatusUnauthorized, gin.H{
					"error": "Unauthorized",
				})
				return
			}
		}

		var req ImmeTranslateRequest

		if err := c.ShouldBindJSON(&req); err != nil {
			c.JSON(http.StatusBadRequest, gin.H{
				"error": err.Error(),
			})
			return
		}

		sourceLang := convertImmeLangToBCP47(req.SourceLang)
		targetLang := convertImmeLangToBCP47(req.TargetLang)

		translations := make([]ImmeTranslation, len(req.TextList))
		ctx, cancel := context.WithTimeout(c.Request.Context(), 120*time.Second)
		defer cancel()

		for i, text := range req.TextList {
			result, err := services.TranslateWithPivot(ctx, sourceLang, targetLang, text, false)
			if err != nil {
				c.JSON(http.StatusInternalServerError, gin.H{
					"error": fmt.Sprintf("Translation failed at index %d: %v", i, err),
				})
				return
			}

			translations[i] = ImmeTranslation{
				DetectedSourceLang: req.SourceLang,
				Text:               result,
			}
		}

		c.JSON(http.StatusOK, ImmeTranslateResponse{
			Translations: translations,
		})
	}
}

File: src/handlers/imme.ts
import { Request, Response } from 'express';
import { normalizeLanguageCode } from '../utils/index.js';
import * as logger from '../logger/index.js';
import { translateWithPivot } from '../services/index.js';

interface ImmeTranslation {
  detected_source_lang: string;
  text: string;
}

export function handleImmeTranslate(apiToken: string) {
  return async (req: Request, res: Response) => {
    if (apiToken) {
      const token = req.query.token as string || '';
      if (token !== apiToken) {
        res.status(401).json({ error: 'Unauthorized' });
        return;
      }
    }

    try {
      const { source_lang, target_lang, text_list } = req.body;

      if (!source_lang || !target_lang || !text_list || !Array.isArray(text_list)) {
        res.status(400).json({ error: 'Missing required fields: source_lang, target_lang, text_list' });
        return;
      }

      const sourceLang = normalizeLanguageCode(source_lang);
      const targetLang = normalizeLanguageCode(target_lang);

      logger.debug(`Imme request: ${sourceLang} -> ${targetLang}, count: ${text_list.length}`);

      const translations: ImmeTranslation[] = [];

      for (let i = 0; i < text_list.length; i++) {
        const text = text_list[i];

        if (!text || typeof text !== 'string') {
          logger.warn(`Skipping invalid text at index ${i}: ${typeof text}`);
          translations.push({
            detected_source_lang: source_lang,
            text: '',
          });
          continue;
        }

        logger.debug(
          `Imme translating [${i + 1}/${text_list.length}]: ${sourceLang} -> ${targetLang}, text length: ${text.length}, preview: ${text.substring(0, 50)}`
        );

        let result: string;
        try {
          result = await translateWithPivot(sourceLang, targetLang, text, false);
          logger.debug(`Imme translated [${i + 1}/${text_list.length}] success`);
        } catch (err) {
          logger.error(`Imme translation failed at index ${i} (${sourceLang} -> ${targetLang}): ${err}`);
          result = text;
        }

        translations.push({
          detected_source_lang: source_lang,
          text: result,
        });
      }

      res.json({ translations });
    } catch (error) {
      logger.error(`Imme translation failed: ${error}`);
      res.status(500).json({ error: `Translation failed: ${error}` });
    }
  };
}
